Add Advanced Recurring Rules to the Event Create/Edit Event modal

bskeen opened the pull request:
I added some UI to the Create and Edit Event Modal to handle advanced recurring events, using the already included RRule library and functionality that was already built into this plugin. Included in these changes:

- Support for Daily, Weekly, Monthly and Yearly events
- A list of supported recurrence types for Monthly and Yearly events
- Automatic conversion from the current `recurring` type to the `rrule` type when the event is opened to edit.
- Some edits to the YAML parsing/serialization logic to support the multiline RRule strings
- Tests for the YAML and recurrence logic
- Works great on desktop and Android (haven't checked iOS)

Please take a look and let me know what you think. I thought this would at least be a good starting point, since this seems to be a highly-requested feature.

NOTE: Support is not yet added in the PR for excluding dates from the rule, but should be fairly easy to add in next. Another feature mentioned in the discussion is the ability to add in your own custom text that will be rendered as a rule. This would also be pretty straightforward to add. I mostly wanted to get something we can start from out there if possible, and then we can improve it going forward.

bskeen committed:
I added some advanced recurring functionality.

bskeen committed:
I added some more tests to cover the different recurrence rules.

bskeen committed:
I removed quotes from the package.json lint commands to make those work cross-platform.

bskeen committed:
I converted some console logs so they would only show up on verbose.

bskeen committed:
I removed one more console log.

MartZOZ commented:
Hey! this is exactly what I was looking for.
I'm quite new to obsidian plugins, is there a way I can implement this myself?

bskeen commented:
Just to clarify, are you asking if there is a way you can get these changes in your local obsidian?

If that is the case, then if you clone the project locally, run `npm i` in the root directory of the project, and then `npm run build`, that should give you the files you need. You just need the `main.js`, `main.css` (which needs to be renamed `styles.css` in order to work), and the manifest.json. Copy those three files into the `.obsidian/plugins/obsidian-full-calendar` directory in your vault, and it should work.

As with any time that you manually install things like this, I would probably suggest taking a backup of your vault just to be sure, although the changes here shouldn't cause any problems. The current recurring events are the only ones that will get updated, and only when they are manually opened/saved. Beyond that, it is just adding new recurring events that should be affected by these changes.

kkartch0 commented:
I tested this on my iPhone SE 2020 (iOS 16.6) with no issues. I tried a variety of options including every n weeks, every n days, every n months, and every n years, all of which worked as I expected them to. Great work!

kkartch0 commented:
If anyone wants to test this pull request without needing to clone down and build, below is a package of the generated plugin files I obtained by building on my machine. You can use this by extracting to the ".obsidian/plugins/obsidian-full-calendar" folder. 

As @bskeen suggested, I also suggest backing up your vault or creating a new test vault when testing plugins. That way you don't risk ruining your primary one. 

[obsidian-full-calendar-advanced-recurrence.zip](https://github.com/davish/obsidian-full-calendar/files/12641934/obsidian-full-calendar-advanced-recurrence.zip)

Mandraw commented:
This seem to work well 👍

davish commented:
This is great, thanks for the awesome work @bskeen! I am posting this from my phone right now but will try and get a full review up in the next day or two. 

As a more meta-point, I'll be trying to work through the backlog of pull requests and new bug reports over the next few weeks. I've been away from OSS for a bit but want to make sure I'm keeping things up to date around here. Cheers!

bskeen committed:
I added a missing class to align the Year/Month select box with the other repeating controls.

bskeen commented:
Thanks, @davish! Maybe once this is merged, I could work on the other features I mentioned, such as excluding dates or custom rules.

davish commented on src/ui/components/event-recurrence-types.test.ts:
```diff
@@ -0,0 +1,2471 @@
+                });
+
+                describe("when day is a positive number", () => {
+                    beforeEach(() => {
```
Just curious: Why write a beforeEach() callback when there is only one test case in this `describe` block?

joeskeen replied:
I'll take responsibility for this as he wrote this according to my testing style. Basically, since the `describe` block defines the condition, semantically I think it makes sense to have all the code connected to that condition in a place where no matter how many tests are in that `describe` context, it is guaranteed to have that condition met. This leaves the `it` statements as one-liners reserved for assertions only. In this case (and admittedly many cases), where you only need to assert one thing, this perhaps becomes more verbose than strictly required; however, it does have the benefit of if you ever need multiple assertions in the future, you can add more `it` blocks and not sacrifice the clear output in case of test failure.

At the end of the day, it is a style choice, just as much as using curly braces for the body of a one-line `if` statement.

For more information on my unit testing philosophy, you can check my talks on the matter (I need to sit down and document it in textual/reference form one of these days).

Conf42 Rustlang 2023: https://youtu.be/-fDIPnG6xJc?si=VA3nY3c2cTDrzZtM (my actual talk, in Rust, but the ideas translate over)
Rendercon Kenya 2023: https://world-class-engineers.github.io/rendercon-2023-unit-testing/ (this is my slides, but the recording isn't available yet)

bskeen replied:
Just to add to what @joeskeen said, I did follow this style; however, I do know that I am contributing to someone else's repository, and so I would be happy to change that if you would like.

davish reviewed (commented):
This is really great work, thanks for the PR! My main request is for you to add some more documentation -- I had some trouble tracing through `event-recurrence-types.ts` and its associated tests, and documentation will make it easier for me and for other contributors to understand this code going forward. Thanks again for you contribution!

bskeen commented on src/ui/components/event-recurrence-types.ts:
```diff
@@ -0,0 +1,402 @@
+    getDisplay: (dateStats: DateStats) => string;
+}
+
+export const MONTH_RECURRENCE_INFO: RecurrenceInfo[] = [
```
The recurrence type is selected inside the MonthYearSelect component, which is in the EditEventRecurrence.tsx file, at the bottom. That provides a select box that is populated with either the monthly or yearly options.

bskeen replied:
Rereading this question, I think you could also mean how is it selected when it is coming in to the component? The answer to that is it looks through the different `RecurrenceInfo`s for the current frequency, and it calls the `hasProps` function on each until one returns true. If they all return false, then it selects the first by default.

bskeen committed:
I added some documentation to the event-recurrence-types file so it is more clear what everything there is doing.

bskeen commented on src/ui/components/event-recurrence-types.ts:
```diff
@@ -0,0 +1,402 @@
+import { DateTime } from "luxon";
```
I added some explanations to this file. Please take a look and see if there is anything else you would like me to document.

bskeen commented on src/ui/components/event-recurrence-types.ts:
```diff
@@ -0,0 +1,402 @@
+    return `${formatOrdinalNumber(value)} to last`;
+};
+
+export interface RecurrenceInfo {
```
I added some explanations to this file. Please take a look and see if there is anything else you would like me to document.

bskeen commented:
Hey @davish I submitted some changes for your feedback, and answered the questions you had. Please let me know if I answered them well enough, or if you need me to change anything else. Thanks again for the great plugin!

superslamram commented:
@davish I just downloaded Full Calendar and bskeen's work would help out alot, can't keep track of birthdays on my own!! I hope it gets merged soon!

bskeen commented:
@davish Just checking in since there hasn't been movement on this in awhile. Is there anything else you need me to fix or add to this PR? Just anxious to begin using it myself.

davish commented:
Hi @bskeen, apologies for my absence from this thread -- life has gotten in the way of open source maintenance recently. The one thing that I would like to see before merging this PR would be to refactor the tests to have them rely less on mutable state between multiple tests in the same suite. It's very difficult for me to follow what the tests are actually asserting, and I think it would make it hard for me to maintain this in the future if any of the tests start failing.

superslamram commented:
@bskeen @davish any updates? alternative solutions?

bskeen committed:
Merge branch 'obsidian-community:main' into main

bskeen committed:
I refactored the recurrence tests to match the requested format.

bskeen committed:
Merge branch 'main' of https://github.com/bskeen/obsidian-full-calendar

bskeen committed:
I added some formatting changes that were mised in the last commit.

bskeen commented:
@davish I finally went back and updated the tests. I removed the shared variables (although I left a few constants, but those don't change between the tests that use them). I also split the large test file into one for each of the different recurrence types. Please take a look and let me know what you think.

bskeen commented:
@davish Just thought I would check-in since it has been a few weeks. It looks like the repo has moved since I opened this PR, are you still the right one to message about this pull request?

matthew-c-lee commented:
@chrisgrieser @joeskeen @joethei @phibr0 @davish 
Is there any reason this hasn't been merged? This is a killer feature for me. I'd be more than happy to take a look at what (if any) problems remain with this MR.

AlgebraicFloat commented:
Amazing stuff!! The only bug I found was creating event titles that were solely numbers (123 for example).

Binaris00 commented:
Good job with this man!! I love it

sago2693 commented:
Please approve the merge!!

VioletJewel commented:
this would be nice to have `rrule` plugin with fullcalendar pretty please.
